Design/ssa-conversion.md

```diff
@@ -0,0 +1,101 @@
+# Converting an ExpEmbedding into SSA form
+
+In order to translate pure Kotlin functions into Viper functions their body must
+be translated into a single expression. Therefore, reassignments of variables may
+not happen on the Viper level which is why the ExpEmbedding must be translated
+into SSA form.
+
+## Converting a Block to SSA form
+
+Assuming we have a block embedding with no branching (branching in the CFG that
+is), converting into SSA form is quite straight-forward:
+- Traverse the ExpressionEmbeddings in the block in order
+- As soon as a reassignment is found introduce a new variable and assign it the
+  RHS of the reassignment.
+- In subsequent usages of the 'original' variable use the newly introduced
+  version.
+In code this may look like the following:
+```kotlin
+var x = 1
+x = x + 1
+return x
+```
+translates to:
+```viper
+let x_0 == (1) in
+let x_1 == (x_0 + 1) in
+x_1
+```
+Note: An optimized encoding of the above example could be created by assigning
+1 + 1 directly to x_0. In general, if no usages occur between two reassignments
+of the same variable we can omit one assignment.
+
+To implement the above algorithm we need to introduce some state tracking the
+assignments of a variable: Then when encountering a variable usage in our block
+we have two options:
+
+1. If the variable is a LHS of an assignment update the variable state to carry
+   the RHS as the new 'latest' version of the variable
```

Comment:
I don't see how this could work; this sounds like what you would do to interpret the code, being a variable to value mapping.  However, the substitution you're doing is a renaming; knowing the binding is not sufficient (though of course you need it *somewhere* to emit it later).